server/src/main/java/org/opensearch/tasks/TaskCancellationMonitoringService.java

```diff
@@ -0,0 +1,202 @@
+            return;
+        }
+        Map<Class<? extends CancellableTask>, List<CancellableTask>> taskCancellationListByType = getCurrentRunningTasksPostCancellation();
+        taskCancellationListByType.forEach((key, value) -> {
+            long uniqueTasksRunningCount = value.stream().filter(task -> {
+                if (this.cancelledTaskTracker.containsKey(task.getId()) && !this.cancelledTaskTracker.get(task.getId())) {
+                    // Mark it as seen by the stats logic.
+                    this.cancelledTaskTracker.put(task.getId(), true);
+                    return true;
+                } else {
+                    return false;
+                }
+            }).count();
+            cancellationStatsHolder.get(key).totalLongRunningCancelledTaskCount.inc(uniqueTasksRunningCount);
+        });
+    }
+
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(() -> {
+            try {
+                doRun();
+            } catch (Exception e) {
+                logger.debug("Exception occurred in Task monitoring service", e);
+            }
+        }, taskCancellationMonitoringSettings.getInterval(), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    protected void doStop() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel();
+        }
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+
+    }
+
+    // For testing
+    protected Map<Long, Boolean> getCancelledTaskTracker() {
+        return this.cancelledTaskTracker;
+    }
+
+    // For testing
+    protected boolean shouldRun() {
+        return shouldRun.get();
+    }
+
+    /**
+     * Invoked when a task is completed. This helps us to disable monitoring service when there are no cancelled tasks
+     * running to avoid wasteful work.
+     * @param task task which got completed.
+     */
+    @Override
+    public void onTaskCompleted(Task task) {
+        if (!TASKS_TO_TRACK.contains(task.getClass())) {
+            return;
+        }
+        if (!this.cancelledTaskTracker.containsKey(task.getId())) {
+            return;
+        }
+        this.cancelledTaskTracker.remove(task.getId());
+        if (this.cancelledTaskTracker.isEmpty()) {
+            shouldRun.set(false);
+        }
+    }
+
+    /**
+     * Invoked when a task is cancelled. This is to keep track of tasks being cancelled. More importantly also helps
+     * us to enable this monitoring service only when needed.
+     * @param task task which got cancelled.
+     */
+    @Override
+    public void onTaskCancelled(CancellableTask task) {
+        if (!TASKS_TO_TRACK.contains(task.getClass())) {
+            return;
+        }
+        // Add task to tracker and mark it as not seen(false) yet by the stats logic.
+        this.cancelledTaskTracker.putIfAbsent(task.getId(), false);
+        if (!shouldRun.get()) {
+            shouldRun.set(true);
+        }
```

Comment:
Great idea of avoiding unnecessary computation from task cancellation monitoring service. Removal from the map and boolean update needs to happen atomically. Else, you might end with non-empty map with shouldRun set to false

Reply:
I see this optimisation to be marred with race conditions at the end we are anyways waiting on thread to be scheduled and checking a boolean condition. Just simplify that to be `cancelledTaskTracker.isEmpty()`.

Reply:
@jainankitk Yeah that's right. Will change.
@Bukhtawar Yeah will simplify to cancelledTaskTracker.isEmpty(), as this boolean now seems unnecessary.